processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -311,21 +306,23 @@ protected IncrementalIndex(
             capabilities,
             dimSchema.getMultiValueHandling()
         );
-        addNewDimension(dimName, capabilities, handler);
+        dimensionData.addNewDimension(dimName, capabilities, handler);
       }
-      columnCapabilities.put(dimName, capabilities);
+      dimensionData.putCapabilities(dimName, capabilities);
     }
 
     //__time capabilities
     ColumnCapabilitiesImpl timeCapabilities = new ColumnCapabilitiesImpl();
     timeCapabilities.setType(ValueType.LONG);
-    columnCapabilities.put(ColumnHolder.TIME_COLUMN_NAME, timeCapabilities);
+    dimensionData.putCapabilities(ColumnHolder.TIME_COLUMN_NAME, timeCapabilities);
 
     // This should really be more generic
     List<SpatialDimensionSchema> spatialDimensions = dimensionsSpec.getSpatialDimensions();
     if (!spatialDimensions.isEmpty()) {
       this.rowTransformers.add(new SpatialDimensionRowTransformer(spatialDimensions));
     }
+
+    this.dimensions = new AtomicReference<>(dimensionData);
```

Comment:
What are the assumptions on the thread-safeness of the constructor? I wonder whether it is 100% safe to make assignment without any synchronization? Do we assume

Reply:
Not sure I understand the question. If the object is not yet constructed, threads shouldn't be able to call any (non-static) methods.